lsp-haskell.el

```diff
@@ -246,15 +246,60 @@ and `lsp-haskell-server-args' and `lsp-haskell-server-wrapper-function'."
 ;; https://microsoft.github.io/language-server-protocol/specification#textDocumentItem
 (add-to-list 'lsp-language-id-configuration '(haskell-literate-mode . "haskell"))
 
+(defun lsp-haskell--github-system-type-suffix ()
+  (pcase system-type
+    ('gnu/linux "Linux")
+    ('windows-nt "Windows")
+    ('darwin "macOS")))
+
+(defcustom lsp-haskell-language-server-github-releases-url
+  "https://github.com/haskell/haskell-language-server/releases/latest/download/"
+  "GitHub releases url for haskell-language-server and its binaries. Make sure to include the trailing directory slash."
+  :type 'string
+  :group 'lsp-haskell
+  :package-version '(lsp-mode . "7.1"))
+
+(lsp-dependency
+ 'haskell-language-server-wrapper
+ '(:system "haskell-language-server-wrapper")
+ `(:download :url ,(concat lsp-haskell-language-server-github-releases-url
+			  (format "haskell-language-server-wrapper-%s.gz"
+				  (lsp-haskell--github-system-type-suffix)))
+	     :store-path ,(f-join lsp-server-install-dir "haskell" "haskell-language-server-wrapper")
+	     :decompress :gzip
+	     :set-executable? t))
+
+(defun lsp-haskell--project-ghc-version ()
+  (let* ((path-exe (executable-find "haskell-language-server-wrapper"))
+	 (downloaded-exe (lsp-package-path 'haskell-language-server-wrapper))
+	 (wrapper-exe (if path-exe path-exe downloaded-exe)))
+    (if wrapper-exe
+      (s-trim
+       (shell-command-to-string
+	(format "%s --project-ghc-version 2>/dev/null" wrapper-exe)))
+      nil)))
+
 ;; Register the client itself
 (lsp-register-client
   (make-lsp--client
-    :new-connection (lsp-stdio-connection (lambda () (lsp-haskell--server-command)))
+   :new-connection (lsp-stdio-connection
+		    (lambda ()
+		      (let* ((proj-ghc-ver (lsp-haskell--project-ghc-version))
```

Comment:
For users who've set up the server binary themselves (and maybe set it with `lsp-haskell-server-path`) it would be nice to not force the use and download of the wrapper.

Reply:
Yeah, this is something that I've yet to figure out, don't take this code as the intended behaviour! I'm wondering how we should decide to do the automatic download: Check for `lsp-haskell-server-path` first and see if it exists?

Reply:
Yeah, something like that makes sense. So `lsp-haskell-server-path` would be how you signal "I know what I'm doing, please just get out of my way and run this binary". And if it's nil, we do the automatic downloading.